src/Orleans/Async/Cancellation/GrainCancellationToken.cs

```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.CodeGeneration;
+using Orleans.Runtime;
+using Orleans.Serialization;
+
+namespace Orleans.Async
+{
+    /// <summary>
+    /// Grain cancellation token that can be passed thought cross-domain boundaries
+    /// </summary>
+    [Serializable]
+    public class GrainCancellationToken : IDisposable
+    {
+        [NonSerialized]
+        private readonly CancellationTokenSource _cancellationTokenSource;
+
+        /// <summary>
+        /// References to remote grains to which this token was passed.
+        /// </summary>
+        [NonSerialized]
+        private readonly ConcurrentDictionary<GrainId, GrainReference> _targetGrainReferences;
+
+        /// <summary>
+        /// Initializes the <see cref="T:Orleans.Async.GrainCancellationToken"/>.
+        /// </summary>
+        internal GrainCancellationToken(
+            Guid id,
+            bool canceled)
+        {
+            _cancellationTokenSource = new CancellationTokenSource();
+            if (canceled)
+            {
+                _cancellationTokenSource.Cancel();
+            }
+
+            Id = id;
+            _targetGrainReferences = new ConcurrentDictionary<GrainId, GrainReference>();
+        }
+
+        /// <summary>
+        /// Unique id of concrete token
+        /// </summary>
+        internal Guid Id { get; private set; }
+
+        /// <summary>
+        /// Underlying cancellation token
+        /// </summary>
+        public CancellationToken CancellationToken => _cancellationTokenSource.Token;
+
+        internal bool IsCancellationRequested => _cancellationTokenSource.IsCancellationRequested;
+
+        internal Task Cancel()
+        {
+            _cancellationTokenSource.Cancel();
+            if (_targetGrainReferences.IsEmpty)
+            {
+                return TaskDone.Done;
+            }
+
+            var cancellationTasks = _targetGrainReferences
+                .Select(pair => pair.Value.AsReference<ICancellationSourcesExtension>()
+                .CancelTokenSource(this))
+                .ToList();
+
+            return Task.WhenAll(cancellationTasks);
+        }
+
+        internal void AddGrainReference(GrainReference grainReference)
+        {
+            _targetGrainReferences.AddOrUpdate(grainReference.GrainId, id => grainReference, (id, ignore) => grainReference);
```

Comment:
why not use `TryAdd` instead of `AddOrUpdate`? it shouldn't matter which instance of GrainReference you use as long as they both reference the same grain, correct?